client/can.go

```diff
@@ -0,0 +1,280 @@
+func NewCanBusClient(nc *nats.Conn, config CanBus) Client {
+	return &CanBusClient{
+		nc:            nc,
+		config:        config,
+		stop:          make(chan struct{}),
+		newPoints:     make(chan NewPoints),
+		newEdgePoints: make(chan NewPoints),
+		wrSeq:         0,
+		lastSendStats: time.Time{},
+		natsSub:       SubjectNodePoints(config.ID),
+	}
+}
+
+// Start runs the main logic for this client and blocks until stopped
+// There are several main aspects of the CAN bus client
+//
+//   - the listener function is a process that recieves CAN bus frames from the Linux
+//	   SocketCAN socket and sends the frames out on the canMsgRx channel
+//
+//   - when a frame is recieved on the canMsgRx channel in the main loop, it is decoded
+//	   and a point is sent out for each canparse.Signal in the frame. The key of each point
+//     contains the message name, signal name, and signal units
+//
+func (cb *CanBusClient) Start() error {
+	log.Println("CanBusClient: Starting CAN bus client: ", cb.config.Description)
+
+	var db *canparse.Database = &canparse.Database{}
+
+	readDb := func() {
+		cb.config.MsgsInDb = 0
+		cb.config.SignalsInDb = 0
+		db.Clean()
+		for _, dbFile := range cb.config.Databases {
+			err := db.ReadBytes([]byte(dbFile.Data), dbFile.Name)
+			if err != nil {
+				log.Println(errors.Wrap(err, "CanBusClient: Error parsing database file"))
+				return
+			} else {
+				/*
+					for _, b := range db.Busses {
+						cb.config.MsgsInDb += len(b.Messages)
+						for _, m := range b.Messages {
+							cb.config.SignalsInDb += len(m.Signals)
+							for _, s := range m.Signals {
+								log.Printf("CanBusClient: read msg %X sig %v: start=%v len=%v scale=%v offset=%v unit=%v",
+									m.Id, s.Name, s.Start, s.Length, s.Scale, s.Offset, s.Unit)
+							}
+						}
+					}
+				*/
+			}
+		}
+		points := data.Points{
+			data.Point{
+				Time:  time.Now(),
+				Type:  data.PointTypeMsgsInDb,
+				Value: float64(cb.config.MsgsInDb),
+			},
+			data.Point{
+				Time:  time.Now(),
+				Type:  data.PointTypeSignalsInDb,
+				Value: float64(cb.config.SignalsInDb),
+			},
+		}
+		err := SendPoints(cb.nc, cb.natsSub, points, false)
+		if err != nil {
+			log.Println(errors.Wrap(err, "CanBusClient: error sending points received from CAN bus: "))
+		}
+	}
+
+	readDb()
+
+	canMsgRx := make(chan can.Frame)
+
+	bringDownDev := func() {}
+
+	setupDev := func() {
+		bringDownDev()
+		conn, err := socketcan.DialContext(context.Background(), "can", cb.config.Device)
+		if err != nil {
+			log.Println(errors.Wrap(err, "CanBusClient: error dialing socketcan context"))
+			return
+		}
+		recv := socketcan.NewReceiver(conn)
+
+		listener := func() {
+			for recv.Receive() {
+				frame := recv.Frame()
+				canMsgRx <- frame
+			}
+		}
+		go listener()
+	}
+
+	setupDev()
+
+	for {
+		select {
+		case <-cb.stop:
+			log.Println("CanBusClient: stopping CAN bus client: ", cb.config.Description)
+			return nil
```

Comment:
the serial driver is not like this, but one pattern I've been using since is to simply break here, and then clean up after the for{} loop. This allows breaking out of the for loop in multiple places if we encounter a critical error that requires the app to be restarted. Not a huge deal -- just FYI. Also, we should call bringDownDev() when client Start() returns.

Reply:
What should I do in the bringDownDev function? The setupDev function uses a socketCan.DialContext call to make the connection.

https://github.com/einride/can-go/blob/8d419f8e19a2414cfd6005fc3b3490aad27bf962/pkg/socketcan/dial.go#L74

Based on the above, when the context sends out a `Done` signal the connection is closed. I don't understand how all this works, so I could use the socketCan.Dial function and close the connection manually in the bringDownDev function.

Reply:
Breaking here only breaks out of the select statement.